src/contrib/tevent.rst

```diff
@@ -0,0 +1,655 @@
+Asynchronous programming: tevent
+################################
+
+SSSD can handle many user requests in parallel. However, the project does not
+use threads to serve the incoming requests but it uses a `talloc`_-based `event
+loop`_ implementation called `tevent`_. Using tevent allows us to neatly avoid
+blocking during communication with remote servers and other processes but at the
+same time we do not have to deal with complicated thread management and locks
+since all the code run in a single thread.
+
+.. seealso::
+
+    The tevent library provides many useful features and mastering it will help
+    you to simplify the code and its logic a lot. However, this document
+    provides only a basic description of tevent and its use cases and patterns
+    inside the SSSD project. It is recommended for you to :tag:`strong` read the
+    official `Tevent Tutorial`_ and `API Reference`_ to learn more about its
+    features. :end-tag:`strong`
+
+.. _event loop: https://en.wikipedia.org/wiki/Event_loop
+.. _talloc: talloc.rst
+.. _tevent: https://tevent.samba.org/
+.. _Tevent Tutorial: https://tevent.samba.org/tevent_tutorial.html
+.. _API Reference: https://tevent.samba.org/group__tevent.html
+
+Event loop
+**********
+
+Event loop is an infinite cycle that awaits an event and executes the registered
+event handler when the event is received. The program goes back to the event
+loop when the handler is done. The event loop uses various mechanisms to receive
+different event types, you may be already familiar with `select`_ or `epoll`_
+that deals with asynchronous input and output operations (read and write to a
+socket).
+
+.. _select: https://man7.org/linux/man-pages/man2/select.2.html
+.. _epoll: https://man7.org/linux/man-pages/man7/epoll.7.html
+
+.. mermaid::
+
+    graph LR
+        loop((Event<br>loop))
+        handler(Event handler)
+        io(Start asynchronous I/O)
+
+        linkStyle default interpolate basis
+        classDef handler fill:green,stroke:green,stroke-width:1px,fill-opacity:0.2
+        style loop fill:red,stroke:red,stroke-width:1px,fill-opacity:0.2,font-weight:bold
+        class handler,io handler
+
+        loop -. await event .-> loop
+        loop -- event received ---> handler
+        handler -- register new event handler ---> io
+        io --> loop
+
+Event types
+===========
+
+The tevent library has support for various event types.
+
+Signals
+    `Signals`_ are a basic IPC mechanism. Tevent provides its own implementation
```

Comment:
I think this link should point to some generic IPC signal description instead of some PR :)

Reply:
This is obviously wrong link. Good catch.